### DISCUSSION TITLE
Why are no detection images ignored in model.track()

### DISCUSSION THREAD

**Discussion by ShivankYadav** [ultralytics/ultralytics#8512] [Q&A] [answered]
I was trying to decouple the tracking functionality of model.track method into detection + processing + tracking. I wanted to apply some basic post processing (handle false positives) to the detection output before applying tracker on it. 
However, when I try to execute the following code:
![image](https://github.com/ultralytics/ultralytics/assets/35888898/d7b3cee8-0246-4b11-97f1-9ed82d7213d1)

I get the following error:
Traceback (most recent call last):
  File "/mnt/tracking.py", line 122, in <module>
    results[i].boxes = Boxes(torch.as_tensor(tracker_out), results[i].orig_shape)
  File "/usr/src/ultralytics/ultralytics/engine/results.py", line 460, in __init__
    assert n in (6, 7), f"expected 6 or 7 values but got {n}"  # xyxy, track_id, conf, cls
AssertionError: expected 6 or 7 values but got 0

It looks like the processing of results object is not expecting empty detections.
Q1) How to resolve this? 

Moreover, I reviewed the "https://github.com/ultralytics/ultralytics/blob/main/ultralytics/trackers/track.py" file's 
"def on_predict_postprocess_end(predictor: object, persist: bool = False) -> None"
method and found out that we are ignoring updating the tracker when there are no detections (line 61).
Q2) Will this not afftect the tracker memory incorrectly, in cases when the person goes out of the frame?

@glenn-jocher Please help.

**Comment by glenn-jocher** [comment, marked as answer]
@ShivankYadav hi there! 👋 It looks like you're diving deep into customizing the tracking process, which is great to see. Let's address your concerns:

### Q1: Handling Empty Detections
The error you're encountering is due to trying to initialize a `Boxes` object with an empty tensor. When there are no detections, the tensor you pass to `Boxes` has a shape of `(0,)`, which doesn't match the expected shape. A quick workaround is to check if `tracker_out` is empty before proceeding:

```python
if tracker_out.shape[0] > 0:
    results[i].boxes = Boxes(torch.as_tensor(tracker_out), results[i].orig_shape)
else:
    results[i].boxes = Boxes(torch.empty((0, 7)), results[i].orig_shape)  # Assuming 7 for xyxy, track_id, conf, cls
```

This way, you're safely handling cases with no detections by explicitly setting an empty `Boxes` object.

### Q2: Tracker Memory and Missing Detections
You're correct in observing that not updating the tracker during frames with no detections could potentially impact tracking continuity, especially in scenarios where objects temporarily leave the frame. The tracker's design assumes continuous input; however, real-world scenarios might indeed involve intermittent detections.

In such cases, it's beneficial for the tracker to have some memory or persistence mechanism to "remember" objects for a certain number of frames without detections. This isn't directly handled in the code snippet you've mentioned but can be implemented as part of the tracker's update logic, depending on the specific tracking algorithm used.

For more nuanced control over tracking, including handling missed detections, you might consider extending the tracker's logic to maintain object IDs for a configurable number of frames without detections. This would involve modifying the tracker's state update mechanism to account for frames where objects are not detected but are expected to reappear based on their last known trajectory.

I hope this helps clarify things! If you have further questions or need more assistance, feel free to ask. Happy coding! 😊

**Reply by ShivankYadav** [reply]
Hi @glenn-jocher Thanks for your input. I was able to do the same after some printing and debugging. 
In terms of tracker, I realized that we are updating the results object (detections) with the tracker output (only for the associated detections) instead of using the tracker output directly. That results in keeping low-confidence false positives as well, which a tracker generally removes. But is there some sort of extra post processing that is happening in or outsite the tracker class becuase I have some output variation in model.track() and detection + tracking even after keeping the same params. 
Regards.

**Reply by glenn-jocher** [reply]
Hi there! 😊 I'm glad to hear you've made progress through some debugging. Regarding your observation on the tracker behavior, you're right; the tracker updates are applied to the detections, which might not always filter out low-confidence detections as one might expect from a tracker alone.

The variation you're seeing between `model.track()` and your custom detection + tracking setup could indeed stem from additional processing steps. While `model.track()` integrates tracking directly with the detection output, including any model-specific post-processing, your custom approach might be bypassing or duplicating some of these steps, depending on how it's implemented.

To ensure consistency, you might want to double-check that:
- All confidence thresholds and non-maximum suppression (NMS) parameters match between the two methods.
- The tracker's initialization and update steps are correctly mirrored in your custom setup.

Here's a quick snippet to ensure NMS and confidence thresholds are aligned:

```python
# Assuming 'detections' is your raw model output
conf_thresh = 0.25
nms_thresh = 0.45
detections = non_max_suppression(detections, conf_thresh, nms_thresh)
```

If you're still noticing discrepancies, it might be helpful to compare the intermediate steps (e.g., pre and post NMS) between `model.track()` and your approach to pinpoint any differences.

Hope this helps clear things up a bit! If you have more questions, feel free to ask. Happy tracking! 🚀

**Reply by ShivankYadav** [reply]
Hi @glenn-jocher. Thanks for your response. I did go through the procedure and from what I understand, I have kept the same params in both model.track() and my decoupled implementation (predict + tracking). Still I not able to figure out the part which may be causing detection inconsistancy in both the output. I will attach the tracking_copy.py file that I am using for both the approaches and the tracker config. Please note that it is just to show the logic and won't run directly.
Moreover, also attached is the ball_botsort.yaml I am using and some extra helper files. 
Please help me if possible. 
[ball_botsort copy.yaml.txt](https://github.com/ultralytics/ultralytics/files/14492862/ball_botsort.copy.yaml.txt)
[tracking_copy.py.txt](https://github.com/ultralytics/ultralytics/files/14492864/tracking_copy.py.txt)
[dataset copy.py.txt](https://github.com/ultralytics/ultralytics/files/14492873/dataset.copy.py.txt)

**Reply by glenn-jocher** [reply]
Hi there! 👋 Thanks for sharing the files and detailing your process. It's great to see you're diving deep into the tracking functionality. Given that you've aligned the parameters across both methods but still observe inconsistencies, it might be worth revisiting a few areas:

1. **Pre- and Post-Processing Steps:** Ensure both approaches use identical pre-processing (e.g., image normalization) and post-processing (e.g., NMS) steps. Sometimes, subtle differences here can lead to varied outputs.

2. **Tracker Initialization:** Double-check that the tracker's initialization parameters match exactly between `model.track()` and your custom setup. Differences in initialization could lead to divergent tracking behavior over time.

3. **Intermediate Outputs:** If possible, log or examine intermediate outputs (detections before and after NMS, tracker states) in both methods. This can help identify where the divergence begins.

Without direct access to the execution environment, it's a bit challenging to pinpoint the exact issue. However, these steps often help isolate and resolve such discrepancies. If you continue facing challenges, feel free to share more details or specific observations that might help us dig deeper. Happy coding! 😊

**Reply by ShivankYadav** [reply]
Hi @glenn-jocher Have been switching states lately, so did not get to close this issue. The thing was that the input videos, although same, were being read by cv2 in model.track() and ffmpeg in my processing script. I got to learn that the frame reading functionality of these encoders are different resulting in different intensity images. Moreover, I found that the YOLO pretrained models work better when input videos/frames are processed with cv2.

**Reply by glenn-jocher** [reply]
@ShivankYadav hi there! 😊 I'm glad to hear you've identified the cause of the discrepancies you were experiencing. It's indeed true that cv2 and ffmpeg might process video streams slightly differently, which can lead to variances in frame quality or color intensity. It's an insightful observation that YOLO pretrained models perform better with cv2-processed input; this could be due to the way cv2 maintains the integrity of the image properties that the models were trained on. Thanks for sharing your findings, and if there's anything more we can help with, please feel free to ask. Happy coding! 🚀
